encryption/src/main/java/org/apache/solr/encryption/EncryptionRequestHandler.java

```diff
@@ -260,12 +339,133 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
+            collectionState = State.ERROR;
+            continue;
+          }
+          encryptRequests.add(() -> sendEncryptionRequestWithRetry(replica, params, solrClient.getClient(), keyId, collectionNameFinal));
+        }
+        try {
+          List<Future<State>> responses = timeAllowedMs <= 0 ?
+              executor.invokeAll(encryptRequests)
+              : executor.invokeAll(encryptRequests, timeAllowedMs, TimeUnit.MILLISECONDS);
+          if (isTimeout(maxTimeNs)) {
+            log.warn("Timeout distributing encryption request for keyId={} collection={}", keyId, collectionName);
+            if (collectionState == null || State.TIMEOUT.priority > collectionState.priority) {
+              collectionState = State.TIMEOUT;
+            }
+          }
+          for (Future<State> response : responses) {
+            State state;
+            try {
+              state = response.get();
+            } catch (ExecutionException e) {
+              collectionState = State.ERROR;
+              break;
+            }
+            if (collectionState == null || state.priority > collectionState.priority) {
+              collectionState = state;
+            }
+          }
+        } catch (InterruptedException e) {
+          collectionState = State.INTERRUPTED;
+        }
+        success = collectionState == null || collectionState.isSuccess();
+      }
+    } finally {
+      String statusValue = success ? STATUS_SUCCESS : STATUS_FAILURE;
+      rsp.add(STATUS, statusValue);
+      rsp.addToLog(STATUS, statusValue);
+      if (collectionState != null) {
+        rsp.add(ENCRYPTION_STATE, collectionState.value);
+        rsp.addToLog(ENCRYPTION_STATE, collectionState.value);
+      }
+      if (log.isInfoEnabled()) {
+        long timeMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNs);
+        log.info("Responding encryption distributed state={} success={} for keyId={} collection={} timeMs={}",
+            (collectionState == null ? null : collectionState.value), success, keyId, collectionName, timeMs);
+      }
+    }
+  }
+
+  private SolrClientHolder getHttpSolrClient(SolrQueryRequest req) {
+    CoreContainer coreContainer = req.getCore().getCoreContainer();
+    CloudSolrClient cloudSolrClient = coreContainer.getSolrClientCache().getCloudSolrClient(coreContainer.getZkController().getZkClient().getZkServerAddress());
+    if (cloudSolrClient instanceof CloudHttp2SolrClient) {
+      return new SolrClientHolder(((CloudHttp2SolrClient) cloudSolrClient).getHttpClient(), false);
+    }
+    return new SolrClientHolder(new Http2SolrClient.Builder().build(), true);
+  }
+
+  protected ModifiableSolrParams createDistributedRequestParams(SolrQueryRequest req, SolrQueryResponse rsp, String keyId) {
+    ModifiableSolrParams params = new ModifiableSolrParams();
+    params.set(PARAM_KEY_ID, keyId == null ? NO_KEY_ID : keyId);
+    return params;
+  }
+
+  boolean isTimeout(long maxTimeNs) {
```

Comment:
this method named confused me a little.  If you name it "hasTimedOut", it'd be a bit clearer.  Also "maxTimeNs" reads looks like it's probably a duration, but it's an instant in time.  "timeoutAtNs" would be clearer.

Reply:
Following your advice, I replaced the code by using TimeOut.
Doing so I discovered a long overflow bug in TimeOut if we try to create a TimeOut with interval = Long.MAX_VALUE, or any large interval which overflows when added to TimeSource.getTimeNs(). I'll create an issue.